stats/opentelemetry/opentelemetry.go

```diff
@@ -0,0 +1,239 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package opentelemetry implements opencensus instrumentation code for gRPC-Go
+// clients and servers.
+package opentelemetry
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal"
+
+	"go.opentelemetry.io/otel/metric"
+)
+
+var logger = grpclog.Component("opentelemetry-instrumentation")
+
+var canonicalString = internal.CanonicalString.(func(codes.Code) string)
+
+var (
+	joinDialOptions = internal.JoinDialOptions.(func(...grpc.DialOption) grpc.DialOption)
+)
+
+// EmptyMetrics represents no metrics. To start from a clean slate if want to
+// pick a subset of metrics, use this and add onto it.
+var EmptyMetrics = Metrics{}
+
+// Metrics is a set of metrics to initialize. Once created, Metrics is
+// immutable.
+type Metrics struct {
+	// metrics are the set of metrics to initialize.
+	metrics map[string]bool
+}
+
+// Add adds the metrics to the metrics set and returns a new copy with the
+// additional metrics.
+func (m *Metrics) Add(metrics ...string) *Metrics {
+	newMetrics := make(map[string]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		newMetrics[metric] = true
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// Remove removes the metrics from the metrics set and returns a new copy with
+// the metrics removed.
+func (m *Metrics) Remove(metrics ...string) *Metrics {
+	newMetrics := make(map[string]bool)
+	for metric := range m.metrics {
+		newMetrics[metric] = true
+	}
+
+	for _, metric := range metrics {
+		delete(newMetrics, metric)
+	}
+	return &Metrics{
+		metrics: newMetrics,
+	}
+}
+
+// MetricsOptions are the metrics options for OpenTelemetry instrumentation.
+type MetricsOptions struct {
+	// MeterProvider is the MeterProvider instance that will be used for access
+	// to Named Meter instances to instrument an application. To enable metrics
+	// collection, set a meter provider. If unset, no metrics will be recorded.
+	// Any implementation knobs (i.e. views, bounds) set in the passed in object
+	// take precedence over the API calls from the interface in this component
+	// (i.e. it will create default views for unset views).
+	MeterProvider metric.MeterProvider
+	// Metrics are the metrics to instrument. Will turn on the corresponding
+	// metric supported by the client and server instrumentation components if
+	// applicable.
+	Metrics Metrics
+	// TargetAttributeFilter is a callback that takes the target string and
+	// returns a bool representing whether to use target as a label value or use
+	// the string "other". If unset, will use the target string as is.
```

Comment:
This is here because the gRFC scopes this out (this was one of the new requirements that came up in between when I wrote this last summer and now - see https://github.com/grpc/proposal/pull/416 for the PR that Yash requested where I add it). This is orthogonal to the concept of target filtration etc. Yes, the canonicalized target string of the channel on the client side is what this filters, which is another dimension to the explicit checks we added, such as the registered method check, for cardinality sakes. This is simply exposing this knob to the user so they have control over this. @yashykt

Reply:
Discussed this offline. Also added language saying "target string of the channel".